util/reward-calculator/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+use ckb_core::header::{BlockNumber, Header};
+use ckb_core::script::Script;
+use ckb_core::transaction::ProposalShortId;
+use ckb_core::Capacity;
+use ckb_store::ChainStore;
+use ckb_traits::ChainProvider;
+use failure::{Error as FailureError, Fail};
+use fnv::FnvHashSet;
+use numext_fixed_hash::H256;
+use std::cmp;
+use std::collections::BTreeMap;
+
+#[derive(Debug, PartialEq, Clone, Eq, Fail)]
+pub enum Error {
+    #[fail(display = "Can't resolve finalize target: {}", _0)]
+    Target(BlockNumber),
+    #[fail(display = "Can't parse Script from target witness: {:x}", _0)]
+    Script(H256),
+}
+
+pub struct RewardCalculator<'a, P> {
+    pub provider: &'a P,
+}
+
+impl<'a, P: ChainProvider> RewardCalculator<'a, P> {
+    pub fn new(provider: &'a P) -> Self {
+        RewardCalculator { provider }
+    }
+
+    pub fn block_reward(&self, parent: &Header) -> Result<(Script, Capacity), FailureError> {
+        let consensus = self.provider.consensus();
+        let store = self.provider.store();
+
+        let block_number = parent.number() + 1;
+        let target_number = consensus
+            .finalize_target(block_number)
+            .ok_or_else(|| Error::Target(block_number))?;
+
+        let target = self
+            .provider
+            .get_ancestor(parent.hash(), target_number)
+            .ok_or_else(|| Error::Target(block_number))?;
+
+        let target_lock = Script::from_witness(
+            &store
+                .get_cellbase(target.hash())
+                .expect("target cellbase exist")
+                .witnesses()[0],
+        )
+        .ok_or_else(|| Error::Script(target.hash().to_owned()))?;
+
+        let txs_fees = self.txs_fees(&target)?;
+        let proposal_reward = self.proposal_reward(parent, &target)?;
+        let base_block_reward = self.base_block_reward(&target)?;
+
+        let reward = txs_fees
+            .safe_add(proposal_reward)?
+            .safe_add(base_block_reward)?;
+        Ok((target_lock, reward))
+    }
+
+    pub fn txs_fees(&self, target: &Header) -> Result<Capacity, FailureError> {
+        let target_ext = self
+            .provider
+            .store()
+            .get_block_ext(target.hash())
+            .expect("block body stored");
+
+        target_ext
+            .txs_fees
+            .iter()
+            .try_fold(Capacity::zero(), |acc, tx_fee| {
+                tx_fee.safe_mul_ratio(4, 10).and_then(|proposer| {
+                    tx_fee
+                        .safe_sub(proposer)
+                        .and_then(|miner| acc.safe_add(miner))
+                })
+            })
+            .map_err(Into::into)
+    }
+
+    pub fn proposal_reward(
```

Comment:
Some notes for other reviewers:

The proposal window is `(2, 10)`.

- target block is `t`
- current block is `t + 11`
- parent block is `t + 10`
- `commit_start` is `t + 2`, which is the earliest block which can commit txs in target block's proposals.
- `proposal_start` is `t - 8`, which is the earliest block which proposals are competing with the target block.

Reply:
Alternatively, the proposed ids can be removed from `target_proposals`, no need to maintain another hash set.